README.md
![alt XOOPS CMS](https://xoops.org/images/logoXoopsPhp8.png)

## uhqgeolocate module for  [XOOPS CMS 2.5.10+](https://xoops.org)

[![XOOPS CMS Module](https://img.shields.io/badge/XOOPS%20CMS-Module-blue.svg)](https://xoops.org)
[![Software License](https://img.shields.io/badge/license-GPL-brightgreen.svg?style=flat)](https://www.gnu.org/licenses/gpl-2.0.html)

[![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/XoopsModules25x/xmodule.svg?style=flat)](https://scrutinizer-ci.com/g/mambax7/xmodule/?branch=master)
[![Latest Pre-Release](https://img.shields.io/github/tag/XoopsModules25x/xmodule.svg?style=flat)](https://github.com/mambax7/xmodule/tags/)
[![Latest Version](https://img.shields.io/github/release/XoopsModules25x/xmodule.svg?style=flat)](https://github.com/mambax7/xmodule/releases/)

**UHQ-GeoLocate** module for [XOOPS CMS](https://xoops.org) is meant to provide geolocation services given an IP address to other XOOPS modules and currently supports the following information services:

* FreeGeoIP (Web API)
* IPInfoDB v2 & v3 (Web API)
* IP2Location (Binary Lookup)
* MaxMind (Web API)

---

This module works on XOOPS 2.5.10. 

Please visit us on https://xoops.org

Current and upcoming "next generation" versions of XOOPS CMS are crafted on GitHub at: https://github.com/XOOPS



docs/changelog.txt
<h5>1.00 Alpha 1   [WORK IN PROGRESS - NOT RELEASED]</h5>  Dev: XOOPS 2.5.11, PHP 7.4.22, PHP 8.0.9
<hr>
- namespaces (mamba)
- class autoloading (mamba)
- renamed 'PHP_SELF' to 'SCRIPT_NAME' (mamba)
- Yoda
- PHP7, PHP8 adjustments (mamba)
- renamed .html to .tpl (mamba)


====[ 0.97 -- 2015-07-19 ]====

Fixes
* Added proper XOOPS root in block template to properly reference the flag locations.  (GitHub-0002 / cesagonchu)
* Added missing error traps for two API providers.  (GitHub-0003 / cesagonchu)

====[ 0.96 -- 2015-06-16 ]====

Code Changes
* Modified directory structure to really fit in GitHub. (Mamba)
* IPInfoDB v2 APIs are depreciated and have been removed as options.  Update scripts will change this accordingly.
* Help section utilizes new help structure for 2.5.7+
* Update script uses language localization in admin.php,
* Update script now produces verbose notes for updates to v0.96.

Fixes
* Removed redundant file check in update script.

Notes
* Dev environment is now XOOPS 2.5.7.1

====[ 0.95 -- Not Released ]====

Fixes
* Fixed a couple rendering issues which would see an error condition.
* Sub-branch adjusting to quirks in XOOPS 2.6.0 Alpha 2.

====[ 0.94 -- 2013-06-15 ]====

Feature Adds
* Ability to cache IPv6 queries w/ API calls.
* MaxMind GeoIP API now supports IPv6 queries.
* IPinfoDB v3 API now supports IPv6 queries.

Code Changes
* Converted to XOOPS 2.5.5 Admin GUI (Mamba)
* Further optimization fo the geolocate.class, code collapses.

Notes
* Changelog back in text format, since it's supported in the Admin GUI.

====[ 0.93 -- 2013-06-11 ]====

Feature Adds
* Added ability to expire cached entries after an arbitrary number of days.
* Added IPInfoDB v3 API support w/ city (IPv4) and country-level lookups (IPv4/IPv6).
* Added FreeGeoIP.net support (IPv4).

Code Changes
* Minor code cleanups and optimizations.
* Removes distribution sample binary files if the module is already installed.

Fixes
* Fixed module upgrade script for v0.91 upgrades.

Notes
* The code repository for the module is now hosted on GitHub under uhqgeolocate for anyone who is interested.  Additionally, the module strings now note the underscore in the module name.

====[ Previous Changes ]====

Full change history is included in the distribution PDF.


xoops_version.php
<?php

// Modular Definitions

defined('XOOPS_ROOT_PATH') || exit('Restricted access');

require_once __DIR__ . '/preloads/autoloader.php';

$moduleDirName = basename(__DIR__);
$moduleDirNameUpper = mb_strtoupper($moduleDirName);

$modversion['version']       = 1.00;
$modversion['module_status'] = 'Beta 1';
$modversion['release_date']  = '2021/08/06';
$modversion['name']          = _MI_UHQGEO_NAME;
$modversion['description']   = _MI_UHQGEO_DESC;
$modversion['author']        = 'Ian A. Underwood';
$modversion['credits']       = 'Underwood Headquarters';
$modversion['help']          = 'page=help';
$modversion['license']       = 'GNU GPL 2.0 or later';
$modversion['license_url']   = 'www.gnu.org/licenses/gpl-2.0.html';
$modversion['official']      = 0;
$modversion['image']         = 'assets/images/logoModule.png';
$modversion['dirname']       = basename(__DIR__);

// Administrative Defines

$modversion['modicons16']          = 'assets/images/icons/16';
$modversion['modicons32']          = 'assets/images/icons/32';
$modversion['module_website_url']  = 'xoops.underwood-hq.org';
$modversion['module_website_name'] = 'XOOPS@UHQ';

// Minimums

$modversion['min_php']   = '7.3';
$modversion['min_xoops'] = '2.5.10';
$modversion['min_admin'] = '1.2';
$modversion['min_db']    = ['mysql' => '5.5'];

// Install/Update Scripts

$modversion['onInstall']   = 'includes/oninstall.php';
$modversion['onUpdate']    = 'includes/onupdate.php';
$modversion['onUninstall'] = 'includes/onuninstall.php';

// Database Information

$modversion['sqlfile']['mysql'] = 'sql/mysql.sql';

$modversion['tables'][] = 'uhqgeolocate_v4cache';
$modversion['tables'][] = 'uhqgeolocate_v6cache';

// Help Section

// ------------------- Help files ------------------- //
$modversion['helpsection'] = [
    ['name' => _MI_UHQGEO_OVERVIEW, 'link' => 'page=help'],
    ['name' => _MI_UHQGEO_DISCLAIMER, 'link' => 'page=disclaimer'],
    ['name' => _MI_UHQGEO_LICENSE, 'link' => 'page=license'],
    ['name' => _MI_UHQGEO_SUPPORT, 'link' => 'page=support'],
    ['name' => _MI_UHQGEO_PREFERENCES, 'link' => 'page=module_prefs'],
];

// Module-Wide Configuration Items

$modversion['hasConfig']   = 1;
$modversion['system_menu'] = 1;

// Option 1: Ability to turn off geolocation site-wide.  When off, query results will all be null.

$modversion['config'][] = [
    'name'        => 'geoloc_ready',
    'title'       => '_MI_UHQGEO_MODCFG_READY',
    'description' => '_MI_UHQGEO_MODCFG_READY_DESC',
    'formtype'    => 'yesno',
    'valuetype'   => 'int',
    'default'     => 1,
];

// Option 2: Select IPv4 DB Provider

$modversion['config'][] = [
    'name'        => 'ipv4_prov',
    'title'       => '_MI_UHQGEO_MODCFG_IPV4',
    'description' => '_MI_UHQGEO_MODCFG_IPV4_DESC',
    'formtype'    => 'select',
    'valuetype'   => 'int',
    'options'     => [
        _MI_UHQGEO_MODCFG_IPV4_DB_IP2L         => 1,
        // IPInfoDB Web API v2 - Depreciated in v0.96
        //_MI_UHQGEO_MODCFG_API_IPDB_CITY_TZ      => 11,
        //_MI_UHQGEO_MODCFG_API_IPDB_CITY         => 12,
        //_MI_UHQGEO_MODCFG_API_IPDB_COUNTRY      => 13,
        // IPInfoDB Web API v3
        _MI_UHQGEO_MODCFG_API_IPDBV3_CITY      => 14,
        _MI_UHQGEO_MODCFG_API_IPDBV3_COUNTRY   => 15,
        // MaxMind Web API
        _MI_UHQGEO_MODCFG_API_MAXMIND_CITY_ISP => 21,
        _MI_UHQGEO_MODCFG_API_MAXMIND_CITY     => 22,
        _MI_UHQGEO_MODCFG_API_MAXMIND_COUNTRY  => 23,
        // FreeGeoIP.net Web API
        _MI_UHQGEO_MODCFG_API_FREEGEOIPNET     => 31,
    ],
    'default'     => 1,
];

// Option 3: Select IPv6 DB Provider

$modversion['config'][] = [
    'name'        => 'ipv6_prov',
    'title'       => '_MI_UHQGEO_MODCFG_IPV6',
    'description' => '_MI_UHQGEO_MODCFG_IPV6_DESC',
    'formtype'    => 'select',
    'valuetype'   => 'int',
    'options'     => [
        _MI_UHQGEO_MODCFG_IPV6_DB_IP2L         => 1,
        // IPInfoDB Web API
        _MI_UHQGEO_MODCFG_API_IPDBV3_COUNTRY   => 15,
        // MaxMind Web API
        _MI_UHQGEO_MODCFG_API_MAXMIND_CITY_ISP => 21,
        _MI_UHQGEO_MODCFG_API_MAXMIND_CITY     => 22,
        _MI_UHQGEO_MODCFG_API_MAXMIND_COUNTRY  => 23,
    ],
    'default'     => 1,
];

// Option 4: Print Data Array on Admin Index

$modversion['config'][] = [
    'name'        => 'geoloc_printr',
    'title'       => '_MI_UHQGEO_MODCFG_PRINTR',
    'description' => '_MI_UHQGEO_MODCFG_PRINTR_DESC',
    'formtype'    => 'yesno',
    'valuetype'   => 'int',
    'default'     => 0,
];

// Option 5: v4 API Key (if Required)

$modversion['config'][] = [
    'name'        => 'geoloc_apikey',
    'title'       => '_MI_UHQGEO_MODCFG_APIV4KEY',
    'description' => '_MI_UHQGEO_MODCFG_APIV4KEY_DESC',
    'formtype'    => 'textbox',
    'valuetype'   => 'text',
    'default'     => '',
];

// Option 6: v6 API Key (if Required)

$modversion['config'][] = [
    'name'        => 'geoloc_apikey_v6',
    'title'       => '_MI_UHQGEO_MODCFG_APIV6KEY',
    'description' => '_MI_UHQGEO_MODCFG_APIV6KEY_DESC',
    'formtype'    => 'textbox',
    'valuetype'   => 'text',
    'default'     => '',
];

// Option 7: Activate API Cache?

$modversion['config'][] = [
    'name'        => 'geoloc_cache',
    'title'       => '_MI_UHQGEO_MODCFG_CACHE',
    'description' => '_MI_UHQGEO_MODCFG_CACHE_DESC',
    'formtype'    => 'yesno',
    'valuetype'   => 'int',
    'default'     => 1,
];

// Option 8: API Cache Expiry

$modversion['config'][] = [
    'name'        => 'geoloc_cacheexpire',
    'title'       => '_MI_UHQGEO_MODCFG_CACHEEXP',
    'description' => '_MI_UHQGEO_MODCFG_CACHEEXP_DEXC',
    'formtype'    => 'int',
    'valuetype'   => 'int',
    'default'     => 0,
];

/**
 * Make Sample button visible?
 */
$modversion['config'][] = [
    'name'        => 'displaySampleButton',
    'title'       => 'CO_' . $moduleDirNameUpper . '_' . 'SHOW_SAMPLE_BUTTON',
    'description' => 'CO_' . $moduleDirNameUpper . '_' . 'SHOW_SAMPLE_BUTTON_DESC',
    'formtype'    => 'yesno',
    'valuetype'   => 'int',
    'default'     => 1,
];

/**
 * Show Developer Tools?
 */
$modversion['config'][] = [
    'name'        => 'displayDeveloperTools',
    'title'       => 'CO_' . $moduleDirNameUpper . '_' . 'SHOW_DEV_TOOLS',
    'description' => 'CO_' . $moduleDirNameUpper . '_' . 'SHOW_DEV_TOOLS_DESC',
    'formtype'    => 'yesno',
    'valuetype'   => 'int',
    'default'     => 0,
];

// Administrative Items

$modversion['hasAdmin']   = 1;
$modversion['adminindex'] = 'admin/index.php';
$modversion['adminmenu']  = 'admin/menu.php';

// Menu Items

$modversion['hasMain'] = 0;

// ------------------- Templates ------------------- //
$modversion['templates'] = [
    ['file' => 'admin/uhqgeolocate_index.tpl', 'description' => _MI_UHQGEO_TEMPLATE_INDEX],
];

// ------------------- Blocks ------------------- //

$i                                       = 0;
$modversion['blocks'][$i]['file']        = 'uhqgeo_blocks.php';
$modversion['blocks'][$i]['name']        = _MI_UHQGEO_BLOCK_FROM_NAME;
$modversion['blocks'][$i]['description'] = _MI_UHQGEO_BLOCK_FROM_DESC;
$modversion['blocks'][$i]['show_func']   = 'b_uhqgeo_from_show';
//$modversion['blocks'][$i]['edit_func']    = "b_uhqgeo_from_edit";
$modversion['blocks'][$i]['template'] = 'uhqgeo_from.tpl';
$modversion['blocks'][$i]['options']  = '';
